### PR TITLE
[FLINK-36517][cdc-connect][paimon] use filterAndCommit API for Avoid commit the same datafile duplicate

### PR DESCRIPTION
**Pull request by beryllw**
https://issues.apache.org/jira/browse/FLINK-35938 problem still persists.

`storeMultiCommitter.commit` API may cause the same datafile commit twice when job restart from failure.

### CODE REVIEW

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonCommitter.java [hunk 1]
```diff
@@ -60,28 +62,12 @@ public void commit(Collection<CommitRequest<MultiTableCommittable>> commitReques
...
-            } else {
-                storeMultiCommitter.commit(Collections.singletonList(wrappedManifestCommittable));
-            }
-            commitRequests.forEach(CommitRequest::signalAlreadyCommitted);
-            LOGGER.info(
-                    String.format(
-                            "Commit succeeded for %s with %s committable",
-                            checkpointId, committables.size()));
-        } catch (Exception e) {
-            commitRequests.forEach(CommitRequest::retryLater);
```
Comment (beryllw) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonCommitter.java:79]: Is there a specific purpose for retrying later in this context?

**Comment by beryllw** [comment]
Could you please assist in reviewing this PR? Thank you. @lvyanquan

**Comment by lvyanquan** [comment]
I agree that the issue of duplicate submissions still exists. Our testing in the case of abnormal failover is relatively lacking, can you try adding corresponding test case for this?

**Comment by beryllw** [comment]
> I agree that the issue of duplicate commits still exists. Our testing in the case of abnormal failover is relatively lacking, can you try adding corresponding test case for this?

I will try, thanks.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/test/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonSinkITCase.java [hunk 2]
```diff
@@ -454,6 +454,46 @@ public void testSinkWithMultiTables(String metastore)
...
+                writer.prepareCommit().stream()
+                        .map(MockCommitRequestImpl::new)
+                        .collect(Collectors.toList());
+        committer.commit(commitRequests);
+
+        // We've two steps in checkpoint: 1. snapshotState(ckp); 2. notifyCheckpointComplete(ckp).
+        // It's possible that flink job will restore from a checkpoint with only step#1 finished and
+        // step#2 not.
+        // CommitterOperator will try to re-commit recovered transactions.
+        committer.commit(commitRequests);
```
[the remark this answers is not in the document]
Comment (lvyanquan) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/test/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonSinkITCase.java:484]: Considering there is another issue https://issues.apache.org/jira/browse/FLINK-36541 in PaimonWriter, If there is a problem with adding this loop, you can skip it for now.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/test/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonSinkITCase.java [hunk 3]
```diff
@@ -454,6 +454,94 @@ public void testSinkWithMultiTables(String metastore)
...
+        }
+        writer.flush(false);
+        Collection<Committer.CommitRequest<MultiTableCommittable>> commitRequests =
+                writer.prepareCommit().stream()
+                        .map(MockCommitRequestImpl::new)
+                        .collect(Collectors.toList());
+        committer.commit(commitRequests);
+
+        // We add a loop for restore 3 times
+        for (int i = 0; i < 3; i++) {
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/test/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonSinkITCase.java:481]: Suggest that the repetition count be greater than 5 to trigger some compaction. 
Refer to [num-sorted-run.compaction-trigger](https://paimon.apache.org/docs/master/maintenance/configurations/#coreoptions) in doc.

**Review by lvyanquan — COMMENTED** [review]
Thanks for the update, and it looks good to me. There is only one small suggestion left.

**Review by lvyanquan — APPROVED** [review]
LGTM.

**Review by leonardBang — APPROVED** [review]
Thanks @beryllw and @lvyanquan for the contribution, +1 and wait the CI green

**Comment by beryllw** [comment]
The test fails because the checkpointId is always 1. I'll fix it.
https://github.com/apache/flink-cdc/pull/3652/files

**Comment by leonardBang** [comment]
@beryllw Would you like to backport this fix to release-3.2 branch?

**Comment by beryllw** [comment]
OK.

**Pull request merged**